### ISSUE TITLE
steering call execution parallelism

### ISSUE THREAD

**Issue by drdozer** [openwdl/wdl#332]
Some calls/tasks have SLA-imposed rate limits. For example, the NCBI eutils have a limit of 3 requests per second, so any workflow that hits these URLs needs to not start a new task instance more frequently than 3 times a second.

Would it be possible to add an additional stansa within a call or a task (whichever makes sense) to steer the parallelism? To sketch a syntax, to illustrate the principle:

```
call task_efetch {
  rateLimit: "3 per second"
  ...
}
```

Some other tasks have a parallelism limit:

```
call task_localBlasN {
  concurrencyLimit: 6
  ...
}
```
maximumTaskStartRate

I understand that under the hood, cromwell uses akka plumbing. I think it is in principle fairly easy to convince the actors or reactive streams to respect this.

**Comment by illusional** [comment]
_Currently_ the [runtime block](https://github.com/openwdl/wdl/blob/master/versions/development/SPEC.md#runtime-section) is probably what you want, technically it's a k-v store so you can put any fields in there you like. 

**However, this is changing very soon: openwdl/wdl#315**, and would become a `hint` you could add that your engine then _may_ respect.

Tbh, I think this is more of an engine thing than WDL, as the WDL spec currently _does_ technically support what you're asking. Cromwell does support global job-rate limiting through the `backend.providers.YourBackend.config.concurrent-job-limit` ([[1](https://github.com/broadinstitute/cromwell/blob/cea07d69919a609362d2e374888f9ed8c4220564/cromwell.examples.conf#L271)] | [[2](https://gatkforums.broadinstitute.org/wdl/discussion/12126/can-cromwell-limit-the-number-of-concurrent-jobs-for-a-particular-task)]) or `system.io.number-of-requests` (they relate to GCP Pipelines API which is similar) [docs/configuring](https://cromwell.readthedocs.io/en/stable/Configuring/).

**Comment by drdozer** [comment]
> Cromwell does support global job-rate limiting through ...

I had read those as global settings, not as things affecting a specific task or call. This is really something that you need to control at this fine-grain level.

**Comment by patmagee** [comment]
While I understand the reason for this, I would argue that this is more an implimentation detail of the individual workflow engine and not in the purview of the WDL specification. The best place for a feature like this would be in the `hints` section which is unstructured KV but does not enforce a contract on the underlying implementation detail.

Rate limits are something that is a bit hard to enforce for a few reasons. First WDL is very unopinionated about what the `scatter` definition measn from an implementation perspective. A scatter call does not in fact guarantee parallel execution, only that the tasks CAN be run in parallel. Nor are there any guarantees about the rate at which jobs within a scatter are launched. These specific details are intentionally left vague because the actual Workflow Engine may or may not be able to implement them. For example, I know cromwell running in local mode does not honor scatter blocks, but instead linearizes the call stack.

Secondly, most tasks do not start right away becuase there is steps to create VM's, localize resources etc. This can take anywhere from seconds to over an hour in some circumstances. In this case a rateLimit does not make much sense. a concurrency limit might be more relevent, but even that is hard to enforce as not every workflow execution engine is sophisticated enough handle limitting scatters.

TBH I am of the opinion that the contract for a scatter should be that each call can run in parrell as defined by the scatter block. Any additional limits and control flow should not be the responsibility of WDL. if a task cannot be run concurrnelty more then `N` times the scatter should likely be created to not be wider then `N`

I could imagine finer grained controls then this, but it likely will need to be a specific engine's implementation of a scatter, using the `hints` section of a task for the control flow

**Comment by BJWiley233** [comment]
Would someone be able to explain if the `backend.providers.YourBackend.config.concurrent-job-limit` actually applies to the scatter method.  I am trying to align and complete alignment QC on 114 fastq files to a reference and so I do this with scatter as in below but my HPC admins indicate the CPU usage increases substantially even when I set `concurrent-job-limit=10`

my.conf
```
include required(classpath("application"))

call-caching {
  enabled = true
}

backend {
	providers {
		BackendName {
			actor-factory = "cromwell.backend.impl.sfs.config.ConfigBackendLifecycleActorFactory"
			config {
				concurrent-job-limit = 10
			}
		}
	}
}
```

workflow where samples is 114 sample structs with fastqs
```
scatter (s in samples) {
		call bwa_task.Mem as bwa {
			input :
				trim = trim,
				read1 = s.read1,
				#read2 = s.read2,
				bwaIndex = bwaIndex,
				outputPrefix = s.outputPrefix,
				readgroup = s.readgroup,
				runtime_params = standard_runtime_bwa	
		}
		
		
		call samtools.sort as samsort {
			input :
				sam = bwa.outputSam,
				outputPrefix = s.outputPrefix,
				runtime_params = standard_runtime_samtools
		}
```

**Comment by illusional** [comment]
We found having call caching on in Cromwell blew up the CPU and disk access. I can’t answer your question directly, but something I’d seen before.

**Comment by BJWiley233** [comment]
Ok thanks, I will wait for someone at Broad to chime in since I am not allowed to run pipeline again until there is a definitive answer or that I am using scatter wrong.  If call-caching is not enabled the entire pipeline has to rerun which sounds like a waste if half of my alignments already completed.

**Comment by patmagee** [comment]
@BJWiley233 if you are waiting for someone from the broad to answer this is not likely the place you are looking for. The specific bwhaviour you are asking about is not an inherent feature of wdl, but implementation specific. I'd suggest checking out the [cromwell github page](https://github.com/broadinstitute/cromwell)` for some answers.
